### PR TITLE
Rework test/run_tests.pl to support selective verbosity and TAP copy

### PR DESCRIPTION
**Pull request by levitte**
This includes a complete rework of how we use `TAP::Harness`, by adding
a `TAP::Parser` subclass that allows additional callbacks to be passed
to perform what we need.  The `TAP::Parser` callbacks we add are:

-   ALL
    to print all the TAP output to a file (conditionally)
    to collect all the TAP output to an array (conditionally)
-   EOF
    to print all the collected TAP output (if there is any) if any subtest failed

To get TAP output to file, the environment variable `HARNESS_TAP_COPY`
must be defined, with a file name as value.  That file will be
overwritten unconditionally.

To get TAP output displayed on failure, the environment variable
`HARNESS_VERBOSE_FAILURE` must be defined with a non-emoty value.

Additionally, the output of test recipe names has been changed to only
display its basename.

### CODE REVIEW

**Comment by levitte** [comment]
This also includes Travis modifications to only get verbose test output on failure.

**Comment by levitte** [comment]
This is an alternative to #9854

**Comment by levitte** [comment]
For demonstration purposes, I added a test recipe that always fails.  *That commit should not be merged*

**Comment by richsalz** [comment]
Impressive!  This is way cool!

**Comment by levitte** [comment]
Looking at one of the Travis jobs:

```
( cd test; \

  mkdir -p test-runs; \

  SRCTOP=../../_srcdist \

  BLDTOP=../. \

  RESULT_D=test-runs \

  PERL="/usr/bin/perl" \

  EXE_EXT= \

  OPENSSL_ENGINES=`cd .././engines 2>/dev/null && pwd` \

  OPENSSL_MODULES=`cd .././providers 2>/dev/null && pwd` \

  OPENSSL_DEBUG_MEMORY=on \

    /usr/bin/perl ../../_srcdist/test/run_tests.pl  )

00-wrong.t ......................... 1/1 

1..1

# text

version: Option unknown option -x

version: Use -help for summary.

/home/travis/build/openssl/openssl/_build/util/shlib_wrap.sh /home/travis/build/openssl/openssl/_build/apps/openssl version -x => 1

not ok 1 - running openssl version -x

#   Failed test 'running openssl version -x'

#   at ../../_srcdist/test/recipes/00-wrong.t line 8.

# Looks like you failed 1 test of 1.

00-wrong.t ......................... Dubious, test returned 1 (wstat 256, 0x100)

Failed 1/1 subtests 

01-test_abort.t .................... ok

01-test_sanity.t ................... ok

01-test_symbol_presence.t .......... ok

01-test_test.t ..................... ok

02-test_errstr.t ................... ok

02-test_internal_context.t ......... ok

02-test_internal_ctype.t ........... ok

02-test_internal_provider.t ........ ok

02-test_lhash.t .................... ok

02-test_ordinals.t ................. ok

02-test_sparse_array.t ............. ok

02-test_stack.t .................... ok

03-test_exdata.t ................... ok

03-test_fipsinstall.t .............. ok

03-test_internal_asn1.t ............ ok

03-test_internal_asn1_dsa.t ........ ok

03-test_internal_bn.t .............. ok

03-test_internal_chacha.t .......... ok

03-test_internal_curve448.t ........ ok

03-test_internal_ec.t .............. ok

03-test_internal_mdc2.t ............ ok

03-test_internal_modes.t ........... ok

03-test_internal_namemap.t ......... ok

03-test_internal_poly1305.t ........ ok

03-test_internal_rsa_sp800_56b.t ... ok

03-test_internal_siphash.t ......... ok

03-test_internal_sm2.t ............. ok

03-test_internal_sm4.t ............. ok

03-test_internal_ssl_cert_table.t .. ok

03-test_internal_x509.t ............ ok

03-test_params_api.t ............... ok

03-test_property.t ................. ok

03-test_ui.t ....................... ok

04-test_asn1_decode.t .............. ok

04-test_asn1_encode.t .............. ok

04-test_asn1_string_table.t ........ ok

04-test_bio_callback.t ............. ok

04-test_bioprint.t ................. ok

04-test_err.t ...................... ok

04-test_param_build.t .............. ok

04-test_params.t ................... ok

04-test_params_conversion.t ........ ok

04-test_pem.t ...................... ok

04-test_provider.t ................. ok

05-test_bf.t ....................... ok

05-test_cast.t ..................... ok

05-test_des.t ...................... ok

05-test_hmac.t ..................... ok

05-test_idea.t ..................... ok

05-test_md2.t ...................... skipped: md2 is not supported by this OpenSSL build

05-test_mdc2.t ..................... ok

05-test_rand.t ..................... ok

05-test_rc2.t ...................... ok

05-test_rc4.t ...................... ok

05-test_rc5.t ...................... skipped: rc5 is not supported by this OpenSSL build

06-test-rdrand.t ................... ok

10-test_bn.t ....................... ok

10-test_exp.t ...................... ok

15-test_dh.t ....................... ok

15-test_dsa.t ...................... ok

15-test_ec.t ....................... ok

15-test_ecdsa.t .................... ok

15-test_ecparam.t .................. ok

15-test_genrsa.t ................... ok

15-test_mp_rsa.t ................... ok

15-test_out_option.t ............... ok

15-test_rsa.t ...................... ok

15-test_rsapss.t ................... ok

20-test_enc.t ...................... ok

20-test_enc_more.t ................. ok

20-test_kdf.t ...................... ok

20-test_mac.t ...................... ok

20-test_passwd.t ................... ok

20-test_pkeyutl.t .................. ok

20-test_provider.t ................. ok

25-test_crl.t ...................... ok

25-test_d2i.t ...................... ok

25-test_pkcs7.t .................... ok

25-test_req.t ...................... ok

25-test_sid.t ...................... ok

25-test_verify.t ................... ok

25-test_x509.t ..................... ok

30-test_aesgcm.t ................... ok

30-test_afalg.t .................... ok

30-test_engine.t ................... ok

30-test_evp.t ...................... ok

30-test_evp_extra.t ................ ok

30-test_evp_kdf.t .................. ok

30-test_evp_pkey_dparam.t .......... ok

30-test_pbelu.t .................... ok

30-test_pkey_meth.t ................ ok

30-test_pkey_meth_kdf.t ............ ok

40-test_rehash.t ................... ok

60-test_x509_check_cert_pkey.t ..... ok

60-test_x509_dup_cert.t ............ ok

60-test_x509_store.t ............... ok

60-test_x509_time.t ................ ok

70-test_asyncio.t .................. ok

70-test_bad_dtls.t ................. ok

70-test_clienthello.t .............. ok

70-test_comp.t ..................... ok

70-test_key_share.t ................ ok

70-test_packet.t ................... ok

70-test_recordlen.t ................ ok

70-test_renegotiation.t ............ ok

70-test_servername.t ............... ok

70-test_sslcbcpadding.t ............ ok

70-test_sslcertstatus.t ............ ok

70-test_sslextension.t ............. ok

70-test_sslmessages.t .............. ok

70-test_sslrecords.t ............... ok

70-test_sslsessiontick.t ........... ok

70-test_sslsigalgs.t ............... ok

70-test_sslsignature.t ............. ok

70-test_sslskewith0p.t ............. ok

70-test_sslversions.t .............. ok

70-test_sslvertol.t ................ ok

70-test_tls13alerts.t .............. ok

70-test_tls13cookie.t .............. ok

70-test_tls13downgrade.t ........... ok

70-test_tls13hrr.t ................. ok

70-test_tls13kexmodes.t ............ ok

70-test_tls13messages.t ............ ok

70-test_tls13psk.t ................. ok

70-test_tlsextms.t ................. ok

70-test_verify_extra.t ............. ok

70-test_wpacket.t .................. ok

80-test_ca.t ....................... ok

80-test_cipherbytes.t .............. ok

80-test_cipherlist.t ............... ok

80-test_ciphername.t ............... ok

80-test_cms.t ...................... ok

80-test_cmsapi.t ................... ok

80-test_ct.t ....................... ok

80-test_dane.t ..................... ok

80-test_dtls.t ..................... ok

80-test_dtls_mtu.t ................. ok

80-test_dtlsv1listen.t ............. ok

80-test_ocsp.t ..................... ok

80-test_pkcs12.t ................... ok

80-test_ssl_new.t .................. ok

80-test_ssl_old.t .................. ok

80-test_ssl_test_ctx.t ............. ok

80-test_sslcorrupt.t ............... ok

80-test_tsa.t ...................... ok

80-test_x509aux.t .................. ok

90-test_asn1_time.t ................ ok

90-test_async.t .................... ok

90-test_bio_enc.t .................. ok

90-test_bio_memleak.t .............. ok

90-test_constant_time.t ............ ok

90-test_fatalerr.t ................. ok

90-test_gmdiff.t ................... ok

90-test_gost.t ..................... skipped: No test GOST engine found

90-test_ige.t ...................... ok

90-test_includes.t ................. ok

90-test_memleak.t .................. ok

90-test_overhead.t ................. skipped: Only supported in no-shared builds

90-test_secmem.t ................... ok

90-test_shlibload.t ................ ok

90-test_srp.t ...................... ok

90-test_sslapi.t ................... ok

90-test_sslbuffers.t ............... ok

90-test_store.t .................... ok

90-test_sysdefault.t ............... ok

90-test_threads.t .................. ok

90-test_time_offset.t .............. ok

90-test_tls13ccs.t ................. ok

90-test_tls13encryption.t .......... ok

90-test_tls13secrets.t ............. ok

90-test_v3name.t ................... ok

95-test_external_boringssl.t ....... skipped: No external tests in this configuration

95-test_external_krb5.t ............ skipped: No external tests in this configuration

95-test_external_pyca.t ............ skipped: No external tests in this configuration

99-test_ecstress.t ................. ok

99-test_fuzz.t ..................... ok

Test Summary Report

-------------------

00-wrong.t                       (Wstat: 256 Tests: 1 Failed: 1)

  Failed test:  1

  Non-zero exit status: 1

Files=177, Tests=1663, 78 wallclock secs ( 3.85 usr  0.35 sys + 68.76 cusr  9.50 csys = 82.46 CPU)

Result: FAIL
```

This is exactly what @richsalz asked for, I think

**Comment by richsalz** [comment]
> This is exactly what @richsalz asked for, I think

Yes. This is awesome.  I really think this, coupled with the ```travis_terminate``` changes will make reviewing CI results painless.  Thank you.

**Comment by levitte** [comment]
I had forgotten to tweak appveyor.yml...

**Comment by levitte** [comment]
Okie, functionality confirmed on Travis and Appveyor.  That makes it time to remove the temporary commit (the one that adds `test/recipes/00-wrong.t`)

**Comment by levitte** [comment]
I squashed and rebased to a fresher master while I was at it.

Time for final review

**Comment by mspncp** [comment]
> To get TAP output displayed on failure, the environment variable
HARNESS_VERBOSE_FAILURE must be defined with a non-emoty value.

Why not just make the default behavior? Why does it require such a long argument to enable such a useful feature?

**Comment by levitte** [comment]
`HARNESS_VERBOSE` is already a known environment variable in [`Test::Harness`](https://perldoc.perl.org/Test/Harness.html#ENVIRONMENT-VARIABLES-THAT-AFFECT-TEST%3a%3aHARNESS) which we have made use of for its documented purpose, so we need to make another.

**Comment by mspncp** [comment]
Then maybe as a compromise add short make parameter? Like `VF=1` for enabling 'verbose on failure'? So we can write

```
nmake VF=1 tests 
```
instead of 
```
nmake HARNESS_VERBOSE_FAILURE=yes tests 
```

**Comment by mspncp** [comment]
But thinking once more about it: why can't  `make tests` just add `HARNESS_VERBOSE_FAILURE=yes` by default? Wouldn't this be the optimal behaviour? Then we don't have to ask users which report test failures to rerun specific tests with `V=1`.

**Comment by mspncp** [comment]
_As for your review request: I will do some usability tests as soon as I find the time, but I doubt that my perl-fu is already sufficient to review the code thoroughly._

**Comment by richsalz** [comment]
There is nobody on the project who can do anything other than trivial review of Richard's perl code.

**Comment by levitte** [comment]
I like the `VF=1` idea. I dislike it being a default, 'cause you can be sure that *someone* will ask "how the he'll do I turn it off???"

**Comment by mspncp** [comment]
> someone will ask "how the he'll do I turn it off???"

This argument does not really convince me yet. If a test fails, who wouldn't like to know _why_ it fails? Just think of all the times you have to instruct people on GitHub to rerun the test using `make V=1 TESTS=...  test`?  Making verbose-on-failure the default would safe us one roundtrip.

 And for those preferring the old behaviour, you could add a silent flag (something like `S=1` or `SILENT=true`).

File: util/perl/OpenSSL/Test.pm [hunk 1]
```diff
@@ -446,16 +446,21 @@ sub run {
...
+    # We're not yet sure this works properly everywhere, so we keep
+    # the code until we've experimented enough to be satisfied
     my $save_STDOUT;
     my $save_STDERR;
-    if ($ENV{HARNESS_ACTIVE} && !$ENV{HARNESS_VERBOSE}) {
-        open $save_STDOUT, '>&', \*STDOUT or die "Can't dup STDOUT: $!";
-        open $save_STDERR, '>&', \*STDERR or die "Can't dup STDERR: $!";
-        open STDOUT, ">", devnull();
-        open STDERR, ">", devnull();
+    if (0) {
```
Comment (t-j-h) [inline util/perl/OpenSSL/Test.pm]: This should not remain - if(0) ...

File: util/perl/OpenSSL/Test.pm [hunk 2]
```diff
@@ -489,16 +494,23 @@ sub run {
...
 
-    if ($ENV{HARNESS_ACTIVE} && !$ENV{HARNESS_VERBOSE}) {
-        close STDOUT;
-        close STDERR;
-        open STDOUT, '>&', $save_STDOUT or die "Can't restore STDOUT: $!";
-        open STDERR, '>&', $save_STDERR or die "Can't restore STDERR: $!";
+    # Make stdout/stderr redirection on non-verbosity dead code
+    # We're not yet sure this works properly everywhere, so we keep
+    # the code until we've experimented enough to be satisfied
+    if (0) {
```
Comment (t-j-h) [inline util/perl/OpenSSL/Test.pm]: This should not remain - if(0) ...

File: util/perl/OpenSSL/Test.pm [hunk 3]
```diff
@@ -1244,8 +1256,13 @@ sub __decorate_cmd {
 
     my $display_cmd = "$cmdstr$stdin$stdout$stderr";
 
-    $stderr=" 2> ".$null
-        unless $stderr || !$ENV{HARNESS_ACTIVE} || $ENV{HARNESS_VERBOSE};
+    # Make stdout/stderr redirection on non-verbosity dead code
+    # We're not yet sure this works properly everywhere, so we keep
+    # the code until we've experimented enough to be satisfied
+    if (0) {
```
Comment (t-j-h) [inline util/perl/OpenSSL/Test.pm]: This should not remain - if(0) ... 

I see the comment - but it needs to go - you have git so you can go back to earlier versions.
Leaving "history" of something you might want to perhaps sometime in the future maybe go back to isn't the right move IMHO. Clean it out.

Comment (levitte) [reply util/perl/OpenSSL/Test.pm]: Pardon me for being cautious, but I want to see how this fares on VMS first.  It's way too easy to "forget" and then wonder where the hell the previous code disappeared.

I will remove as soon as I've ensured his doesn't cause trouble there.  So the question is if you want to delay *this PR* on pure principle, or if you can find yourself waiting for a clean-up PR within a week or so?

I would normally just have this PR waiting, but this PR also solves a current Travis problem, so *this time* I can see reason to do that clean-up in a separate PR.

**Comment by levitte** [comment]
> Making verbose-on-failure the default would safe us one roundtrip.

Until it hits an error in one of those huge tests, like test_evp, and someone quotes that last three lines...

**Comment by mspncp** [comment]
You win. I‘m satisfied with VF=1 too.

**Comment by levitte** [comment]
Fixup that adds `VERBOSE_FAILURE` as well as `VF`, plus changes to `INSTALL` and `CHANGES`

File: util/perl/OpenSSL/Test.pm (hunk 3 above)
Comment (levitte) [reply util/perl/OpenSSL/Test.pm]: So I rushed the experiment...  and it turns out that, for reasons I can't understand, some program output isn't caught by TAP::Parser on VMS.  So this code will be re-enabled, but only on VMS.

**Comment by levitte** [comment]
Ping

**Comment by mspncp** [comment]
> Ping

Sigh. How fortunate you are to be able to work 24/7 on OpenSSL...  ;-)

**Comment by levitte** [comment]
> Sigh. How fortunate you are to be able to work 24/7 on OpenSSL... ;-)

What, feeling pressured? ;-)

**Review by paulidale — COMMENTED** [review]
I'll have another look tomorrow when I've a clearer head, but I fear this might be beyond my Perl skillset.

**Review by mattcaswell — APPROVED** [review]
Approved subject to the typo being fixed.

**Pull request closed without merging**